_source/_posts/2022-02-23-angular-async-config.md

```diff
@@ -0,0 +1,340 @@
+
+Angular has a built-in way to support per environment differences using the `environments/environment.*.ts` files. When serving locally, Angular CLI uses the values in `environments/environment.ts`, and when you build for prod, Angular CLI substitutes `environment.prod.ts` instead. You can see this file substitution defined in the `angular.json` build configuration. And if you have more environments to support, you can customize the build configuration to suit your needs.
+
+The environment files are helpful when you have different configurations you want to support at build time. Some examples include enabling user analytics only on prod environments or defining the API endpoints your QA environment calls.
+
+An example of configuration that changes by environment already built into your app is in `src/main.ts`. Here you see the following:
+
+```ts
+import { enableProdMode } from '@angular/core';
+import { environment } from './environments/environment';
+
+if (environment.production) {
+  enableProdMode();
+}
+```
+
+Angular utilizes the environment files to identify when to call the `enableProdMode()` method. Notice the file imports from `./environments/environment`. That's because the build process handles that file swap.
+
+Now let's look at how we'd use this when integrating with Okta. 
+
+In `src/environments/environment.ts`, add the Okta auth configuration like this.
+
+```ts
+export const environment = {
+  production: false,
+  authConfig: {
+    issuer: 'https://{yourOktaDomain}/oauth2/default',
+    clientId: '{yourClientId}'
+  }
+};
+```
+
+In `src/environments/environment.prod.ts`, you'll add the same `authConfig` properties with values that match your prod environment.
+
+You'll use the environment file to initialize the `OktaAuthModule` in the `AppModule` like this.
+
+```ts
+import { OktaAuthModule, OKTA_CONFIG } from '@okta/okta-angular';
+import { OktaAuth } from '@okta/okta-auth-js';
+import { environment } from '../environments/environment.ts';
+
+const oktaAuth = new OktaAuth({
+  ...environment.authConfig,
+  redirectUri: window.location.orgin + '/login/callback'
+});
+
+@NgModule({
+  declarations: [ AppComponent, ProfileComponent ],
+  imports: [ BrowserModule, AppRoutingModule, OktaAuthModule ],
+  providers: [
+    { provide: OKTA_CONFIG, useValue: { oktaAuth }}
+  ],
+  bootstrap: [AppComponent]
+})
+export class AppModule { }
+```
+
+**Summary**
+
+Customizing environment files is the Angular recommended method to inject values during build time.
+
+**When to use**
+
+* You have different configuration values based on build output
+
+**Best for**
+
+* Devmode - keep apps served locally from doing things only prod apps should do 
+* Multiple staging environment systems
+
+**Watch out for**
+
+* Configuration that involves private keys or tokens
+* Run your build for all the different environments when testing changes you make. You don't want to miss adding a property and potentially get a runtime error.
```

Comment:
```suggestion
* Run your build for all your different environments when testing changes you make. You don't want to miss adding a property and potentially get a runtime error.
```
is this more precise?

Reply:
How about 
> Run the build for each environment to test changes you make. You don't want to miss adding a property and potentially get a runtime error.

I'll go with this for now and open to review suggestions

Reply:
👍